### PR TITLE
Header for dpctl sycl interface type casters

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
This PR adds "libsyclinterface/include/dpctl_sycl_type_casters.hpp" which defines 
`dpctl::syclinterface::dpctl_device_selector` class corresponding to `DPCTLDeviceSelectorRef` opaque pointer.

The class is derived from `sycl::device_selector` to support oneAPI DPC++ 2022.2 and older, but is a stand-alone
class if `__SYCL_COMPILER_VERSION >= 20221020L`. 

Use of `dpctl_device_selector` in `sycl::device` and `sycl::platform` constructors is adjusted based on the compiler version too.

This PR also exposes `dpctl::syclinterface::wrap<T>` and `dpctl::syclinterface::unwrap<T>` templated functions where the template type must be the underlying sycl type to the pointer of which the opaque pointer should be `reinterpret_cast`-ed.

For example:

```c++
   DPCTLSyclDeviceRef DRef;
   auto Dev = ::dpctl::syclinterface::unwrap<sycl::device>(DRef);
```

libsyclinterface/source and libsyclinterface/tests are adjusted to use exported wrap/unwrap.

- [x] Have you provided a meaningful PR description?
- [ ] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [ ] Have you made sure that new changes do not introduce compiler warnings?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

**Comment by oleksandr-pavlyk** [comment]
Should this PR also remove setting of `SYCL_ENABLE_HOST_DEVICE=1` in workflows?

File: libsyclinterface/helper/source/dpctl_utils_helper.cpp [hunk 1]
```diff
@@ -242,9 +251,13 @@ std::string DPCTL_AspectToStr(aspect aspectTy)
 aspect DPCTL_StrToAspectType(const std::string &aspectTyStr)
 {
     aspect aspectTy;
-    if (aspectTyStr == "host") {
+    if (false) {
```
Comment (diptorupd) [inline libsyclinterface/helper/source/dpctl_utils_helper.cpp:254]: Why not just move the "cpu" check on top?

File: libsyclinterface/include/dpctl_device_selection.hpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+#include <CL/sycl.hpp>
+
+namespace dpctl
+{
+namespace syclinterface
+{
+
+#if __SYCL_COMPILER_VERSION >= __SYCL_COMPILER_2023_SWITCHOVER
+
+class DPCTL_API dpctl_device_selector
```
Comment (diptorupd) [inline libsyclinterface/include/dpctl_device_selection.hpp:40]: As discussed this class should be in a separate header.

**Review by diptorupd — DISMISSED** [review]
I did not go very deep as the changes are fairly mechanical and we have already discussed them offline.

Please merge.

File: libsyclinterface/helper/source/dpctl_utils_helper.cpp (hunk 1 above)
Comment (oleksandr-pavlyk) [reply libsyclinterface/helper/source/dpctl_utils_helper.cpp:254]: Good point

**Pull request merged**